aptos/contracts/sources/pyth.move

```diff
@@ -0,0 +1,803 @@
+        // Update the price feed from each VAA
+        while (!vector::is_empty(&vaas)) {
+            update_price_feed_from_single_vaa(vector::pop_back(&mut vaas));
+        };
+
+        // Charge the message update fee
+        assert!(get_update_fee() <= coin::value(&fee), error::insufficient_fee());
+        coin::deposit(@pyth, fee);
+    }
+
+    fun update_price_feed_from_single_vaa(vaa: vector<u8>) {
+        // Deserialize the VAA
+        let vaa = vaa::parse_and_verify(vaa);
+
+        // Check that the VAA is from a valid data source (emitter)
+        assert!(
+            state::is_valid_data_source(
+                data_source::new(
+                    u16::to_u64(vaa::get_emitter_chain(&vaa)),
+                    vaa::get_emitter_address(&vaa))),
+            error::invalid_data_source());
+
+        // Deserialize the batch price attestation
+        update_cache(batch_price_attestation::destroy(batch_price_attestation::deserialize(vaa::destroy(vaa))));
+    }
+
+    /// Update the cache with given price updates, if they are newer than the ones currently cached.
+    fun update_cache(updates: vector<PriceInfo>) {
+        while (!vector::is_empty(&updates)) {
+            let update = vector::pop_back(&mut updates);
+            if (is_fresh_update(&update)) {
+                let price_feed = *price_info::get_price_feed(&update);
+                let price_identifier = price_feed::get_price_identifier(&price_feed);
+                state::set_latest_price_info(
+                    *price_identifier,
+                    update,
+                );
+                event::emit_price_feed_update(price_feed, timestamp::now_microseconds());
+            }
+        };
+        vector::destroy_empty(updates);
+    }
+
+    /// Update the cached price feeds with the data in the given VAAs, using
+    /// update_price_feeds(). However, this function will only have an effect if any of the
+    /// prices in the update are fresh. The price_identifiers and publish_times parameters
+    /// are used to determine if the update is fresh without doing any serialisation or verification
+    /// of the VAAs, potentially saving time and gas. If the update contains no fresh data, this function
+    /// will revert with error::no_fresh_data(). 
+    /// 
+    /// For a given price update i in the batch, that price is considered fresh if the current cached 
+    /// price for price_identifiers[i] is older than publish_times[i].
+    public entry fun update_price_feeds_if_fresh(
+        vaas: vector<vector<u8>>,
+        price_identifiers: vector<vector<u8>>,
+        publish_times: vector<u64>,
+        fee: Coin<AptosCoin>) {
+
+        assert!(vector::length(&price_identifiers) == vector::length(&publish_times),
+            error::invalid_publish_times_length());
+
+        let fresh_data = false;
+        let i = 0;
+        while (i < vector::length(&publish_times)) {
+            let price_identifier = price_identifier::from_byte_vec(
```

Comment:
I'm not too familiar with move, but looking at `with_funder` method above, shouldn't this be the same? or have a same `with_funder` version? (For an imaginary price pusher service that wants to send update directly)

Reply:
Yep, I just didn't have time to add this - I will add one in a follow-up PR.